fe/fe-core/src/main/java/com/starrocks/warehouse/Warehouse.java

```diff
@@ -145,199 +76,26 @@ public void setState(WarehouseState state) {
 
-                long clusterId = GlobalStateMgr.getCurrentState().getNextId();
-                Cluster cluster = new Cluster(clusterId, workerGroupId);
-                clusters.put(clusterId, cluster);
-                AlterWhClusterOplog log = new AlterWhClusterOplog(this.getFullName(), cluster);
-                GlobalStateMgr.getCurrentState().getEditLog().logAddCluster(log);
+    public abstract void dropSelf() throws DdlException;
 
-                this.state = WarehouseState.RUNNING;
-            } else {
-                throw new DdlException("cluster num of " + name + "is " +
-                        clusters.size() + " exceed max cluster limit " + maxCluster);
-            }
-        } finally {
-            writeUnLock();
-        }
-    }
-
-    public void removeCluster() throws DdlException {
-        writeLock();
-        try {
-            if (clusters.size() > minCluster) {
-                Long clusterId = clusters.keySet().stream().findFirst().orElse(-1L);
-                Cluster cluster = clusters.get(clusterId);
-                GlobalStateMgr.getCurrentStarOSAgent().deleteWorkerGroup(cluster.getWorkerGroupId());
-                clusters.remove(clusterId);
-                AlterWhClusterOplog log = new AlterWhClusterOplog(this.getFullName(), cluster);
-                GlobalStateMgr.getCurrentState().getEditLog().logRemoveCluster(log);
-            } else {
-                throw new DdlException("cluster num of " + name + " exceed min cluster limit " + minCluster);
-            }
-        } finally {
-            writeUnLock();
-        }
-    }
+    public abstract Map<Long, Cluster> getClusters() throws DdlException;
 
-    public void replayAddCluster(AlterWhClusterOplog log) {
-        writeLock();
-        try {
-            Cluster cluster = log.getCluster();
-            clusters.put(cluster.getId(), cluster);
-            this.state = WarehouseState.RUNNING;
+    public abstract void setClusters(Map<Long, Cluster> clusters) throws DdlException;
 
-        } finally {
-            writeUnLock();
-        }
-    }
-
-    public void replayRemoveCluster(AlterWhClusterOplog log)  {
-        writeLock();
-        try {
-            Cluster cluster = log.getCluster();
-            clusters.remove(cluster.getId());
-
-        } finally {
-            writeUnLock();
-        }
-    }
-
-    public void getProcNodeData(BaseProcResult result) {
-        result.addRow(Lists.newArrayList(this.getFullName(),
-                this.getState().toString(),
-                this.getSize(),
-                String.valueOf(this.getMinCluster()),
-                String.valueOf(this.getMaxCluster()),
-                String.valueOf(this.clusters.size()),
-                String.valueOf(this.getTotalPendingSqls()),
-                String.valueOf(this.getTotalRunningSqls())));
-    }
+    public abstract void addNodes(Map<Long, ComputeNode> cns);
 
     @Override
     public void write(DataOutput out) throws IOException {
```

Comment:
this is an abstract class, is it Okay to have write/read, because you can't construct a warehouse class now.